plugins/c8y_firmware_plugin/src/download.rs

```diff
@@ -0,0 +1,336 @@
+        let firmware_url = smartrest_request.url.as_str();
+        let file_cache_key = digest(firmware_url);
+
+        // <tedge-cache-root>/<file_cache_key>
+        let cache_dest = PathBuf::from(FILE_TRANSFER_CACHE_PATH).join(&file_cache_key);
+        let cache_dest_str = format!("{FILE_TRANSFER_CACHE_PATH}/{file_cache_key}");
+
+        // <tedge-file-transfer-root>/<child-id>/firmware_update/<file_cache_key>
+        let transfer_dest = PathBuf::from(FILE_TRANSFER_ROOT_PATH)
+            .join(child_id)
+            .join("firmware_update")
+            .join(&file_cache_key);
+
+        let file_transfer_url = format!(
+            "http://{}/tedge/file-transfer/{child_id}/firmware_update/{file_cache_key}",
+            &self.local_http_host
+        );
+
+        // If dir already exists, these calls do nothing.
+        create_parent_dirs(&transfer_dest)?;
+        create_parent_dirs(&cache_dest)?;
+
+        if cache_dest.is_file() {
+            info!("Hit the cache={cache_dest_str}. File download is skipped.");
+        } else {
+            match self
+                .http_client
+                .lock()
+                .await
+                .download_file(firmware_url, &file_cache_key, &self.tmp_dir)
+                .await
+            {
+                Ok(tmp_file_path) => {
+                    info!("Successfully downloaded from {firmware_url}");
+                    move_file(&tmp_file_path, &cache_dest)?;
+                }
+                Err(err) => {
+                    error!("Failed to download from {firmware_url}");
+                    return Err(err.into());
+                }
+            }
+        }
+
+        // Create a symlink if it doesn't exist yet.
+        if !transfer_dest.is_file() {
+            unix_fs::symlink(&cache_dest, &transfer_dest)?;
+        }
+
+        // Add a pair of local url and external url to the hashmap.
+        self.url_map
+            .insert(file_transfer_url.clone(), firmware_url.to_string());
+
+        let file_sha256 = try_digest(transfer_dest.as_path())?;
+        let firmware_entry =
+            FirmwareEntry::new(firmware_name, firmware_version, firmware_url, &file_sha256);
+        let firmware_op_req = FirmwareOperationRequest::new(child_id, firmware_entry);
+        let firmware_update_req_msg = Message::new(
+            &firmware_op_req.operation_request_topic(),
+            firmware_op_req.operation_request_payload(file_transfer_url.as_ref())?,
+        );
+
+        self.mqtt_publisher.send(firmware_update_req_msg).await?;
+        info!(
+            "Firmware update request is sent to child device. \
+            Details: child={child_id}, name={firmware_name}, version={firmware_version}, url={file_transfer_url}"
+        );
+
+        // A unique ID for timer. Once operation ID is possible to use, better to replace it.
+        let timer_id = digest(format!(
+            "{child_id}{firmware_name}{firmware_version}{firmware_url}"
```

Comment:
Isn't the `firmware_url` unique enough for a firmware? Do we really need to append the `child_id`, `firmware_name` etc as well to make it "more unique"?

Reply:
Having all 3 fragments (name, version, url) ensures that the timeout is only for this specific information. Just in case if people use the firmware name/version as some additional information which affects how the url is processed (or to use the name/version as re-processing logic, to enforce applying the same firmware binary again, but by just incrementing the version or changing the name).

Reply:
Actually, this was already replaced by `{opration_id}`.